x-pack/legacy/plugins/epm/public/screens/add_data_source/add_data_source_form.tsx

```diff
@@ -56,58 +73,134 @@ export const AddDataSourceForm = ({ package: pkg }: AddDataSourceStepsProps) =>
   }, []);
 
   const [addDataSourceSuccess, setAddDataSourceSuccess] = useState<boolean>(false);
+
   const [datasourceName, setDatasourceName] = useState<FormState['datasourceName']>('');
+  const [datasourceError, setDatasourceError] = useState<FormStateRowErrors['datasourceName']>();
+
   const [selectedDatasets, setSelectedDatasets] = useState<FormState['datasets']>({});
+  const [datasetsError, setDatasetsError] = useState<FormStateRowErrors['datasets']>();
+
   const [selectedPolicies, setSelectedPolicies] = useState<FormState['policies']>([
     defaultPolicyOption,
   ]);
+  const [policiesError, setPoliciesError] = useState<FormStateRowErrors['policies']>();
 
   const formState: FormState = {
     datasourceName,
     datasets: selectedDatasets,
     policies: selectedPolicies,
+    errors: {
+      datasourceName: datasourceError,
+      datasets: datasetsError,
+      policies: policiesError,
+    },
   };
 
   const { notifications } = useCore();
   const { toDetailView } = useLinks();
   const datasets = pkg?.datasets || [];
-  const handleRequestInstallDatasource = async () => {
-    try {
-      await installDatasource({
-        pkgkey: `${pkg.name}-${pkg.version}`,
-        datasets: datasets.filter(d => formState.datasets[d.name] === true),
-        datasourceName: formState.datasourceName,
-        // @ts-ignore not sure where/how to enforce a `value` key on options
-        policyIds: formState.policies.map(({ value }) => value),
-      });
-      setAddDataSourceSuccess(true);
-      notifications.toasts.addSuccess({
-        title: `Added ${pkg.title} data source`,
-      });
-      return;
-    } catch (err) {
-      notifications.toasts.addWarning({
-        title: `Failed to add data source to ${pkg.title}`,
-        iconType: 'alert',
-      });
+  useEffect(() => {
```

Comment:
What is the gain of doing this in an useEffect instead of the `handleSubmit`?

Reply:
It works :)

We want to validate the form before submitting but validation updates state variables which causes a re-render/update. Moving the submission behavior (a side effect) into `useEffect` allows the component to be called multiple times, but only submit when it's in the desired state.

Reply:
This is a very "React" way to do this vs a more standard "js" way of use onChange + onsubmit. I think keeping things as "js" as is reasonable is best. That is to say if you prefer it to work this way that's is fine... but using useEffect would not be the only way to accomplish this

Reply:
I take your point, but this is using [`EuiSteps`](https://elastic.github.io/eui/#/navigation/steps) and some other components which don't generate a `form` element so there's no `submit` event / `onSubmit` handler. The existing component and prior PR use React Hooks so I continued with that here.